src/main/java/sirius/biz/jobs/batch/file/DictionaryBasedImport.java

```diff
@@ -0,0 +1,158 @@
+        this.rowHandler = rowHandler;
+    }
+
+    /**
+     * Invoked for each row in an import file.
+     *
+     * @param index the line number to process
+     * @param row   the data to process
+     */
+    public void handleRow(int index, Values row) {
+        if (index == 1) {
+            handleHeaderRow(row);
+        } else {
+            handleDataRow(index, row);
+        }
+    }
+
+    private void handleHeaderRow(Values row) {
+        if (dictionary.hasIdentityMapping()) {
+            boolean problemsFound = dictionary.detectHeaderProblems(row, (problem, isFatal) -> {
+                if (Boolean.TRUE.equals(isFatal)) {
+                    process.log(ProcessLog.error()
+                                          .withNLSKey(ERROR_IN_ROW_KEY)
+                                          .withContext("row", 1)
+                                          .withContext(MESSAGE_KEY, problem));
+                } else {
+                    process.log(ProcessLog.warn()
+                                          .withNLSKey(ERROR_IN_ROW_KEY)
+                                          .withContext("row", 1)
+                                          .withContext(MESSAGE_KEY, problem));
+                }
+            }, true);
+
+            if (problemsFound) {
+                TaskContext.get().cancel();
+            }
+        } else {
+            dictionary.resetMappings();
+            try {
+                dictionary.determineMappingFromHeadings(row, false);
+                process.log(ProcessLog.info().withMessage(dictionary.getMappingAsString()));
+            } catch (HandledException e) {
+                process.log(ProcessLog.error()
+                                      .withNLSKey(ERROR_IN_ROW_KEY)
+                                      .withContext("row", 1)
+                                      .withContext(MESSAGE_KEY, e.getMessage()));
+                TaskContext.get().cancel();
+            }
+        }
+    }
+
+    private void handleDataRow(int index, Values row) {
+        if (row.length() == 0) {
+            return;
+        }
+
+        Watch watch = Watch.start();
+        try {
+            Context context = filterEmptyValues(dictionary.load(row, false));
+            if (!isEmptyContext(context)) {
+                rowHandler.invoke(Tuple.create(index, context));
+            }
+        } catch (HandledException e) {
+            process.incCounter(NLS.get("LineBasedJob.erroneousRow"));
+            process.handle(Exceptions.createHandled()
+                                     .to(Log.BACKGROUND)
+                                     .withNLSKey(ERROR_IN_ROW_KEY)
+                                     .set("row", index)
+                                     .set(MESSAGE_KEY, e.getMessage())
+                                     .handle());
+        } catch (Exception e) {
+            process.incCounter(NLS.get("LineBasedJob.erroneousRow"));
+            process.handle(Exceptions.handle()
```

Comment:
We could expose a method so jobs can easily create process logs for errors/warning with a custom message like "Product XXX not found", respecting the message format (row number + file name)

Reply:
this is was the handler above (line 128) should do?

Reply:
I mean, a job might want to write in the process log a message like "Error in file my_file, row ##: whatever message", but keep on processing more stuff. (Example with Product not found was not the best I confess). More of a message "Cannot set property x with value y", out of a list of many properties. Or shall the jobs cumulate those and thrown them as a single error?